dbms/src/Interpreters/Aggregator.cpp

```diff
@@ -770,30 +868,44 @@ std::optional<typename Method::template EmplaceOrFindKeyResult<only_lookup>::Res
     }
 }
 
-template <typename Method>
-ALWAYS_INLINE inline void prepareBatch(
+template <bool enable_prefetch, bool batch_get_key_holder, typename Method, typename KeyHolderType>
+ALWAYS_INLINE inline void setupKeyHolderAndHashVal(
     size_t row_idx,
-    size_t end_row,
+    size_t batch_size,
     std::vector<size_t> & hashvals,
-    std::vector<typename Method::State::Derived::KeyHolderType> & key_holders,
+    std::vector<KeyHolderType> & key_holders,
     Arena * aggregates_pool,
     std::vector<String> & sort_key_containers,
     Method & method,
     typename Method::State & state)
 {
-    assert(hashvals.size() == key_holders.size());
+    key_holders.resize(batch_size);
+    if constexpr (enable_prefetch)
+        hashvals.resize(batch_size);
 
-    for (size_t i = row_idx, j = 0; i < row_idx + hashvals.size() && i < end_row; ++i, ++j)
+    for (size_t i = row_idx, j = 0; i < row_idx + batch_size; ++i, ++j)
     {
-        key_holders[j] = static_cast<typename Method::State::Derived *>(&state)->getKeyHolder(
-            i,
-            aggregates_pool,
-            sort_key_containers);
-        hashvals[j] = method.data.hash(keyHolderGetKey(key_holders[j]));
+        if constexpr (batch_get_key_holder)
+            key_holders[j]
+                = static_cast<typename Method::State::Derived *>(&state)->getKeyHolderBatch(j, aggregates_pool);
+        else
+            key_holders[j] = static_cast<typename Method::State::Derived *>(&state)->getKeyHolder(
+                i,
+                aggregates_pool,
+                sort_key_containers);
+
+        if constexpr (enable_prefetch)
```

Comment:
Is it expected that hashvals not updated when enable_prefetch is false?

Reply:
Yes. When enable_prefetch is false and batch_get_key_holder is true, will use `emplaceKey(method, state, key_holder)` to emplace into HashTable. So no need to init hashvals

Reply:
hashvals is only for prefetch to use. so no need to compute hashval for other situations

Reply:
ok